README.md
# cabot-cookbook
![travis build status](https://travis-ci.org/rafaelmagu/chef-cabot.svg?branch=master)


[Cabot](http://cabotapp.com) is a self-hosted watchdog for your websites and infrastructure. This cookbook installs and configures a basic setup.

## Supported Platforms

Made with love for Ubuntu (12.04, 14.04)

## Dependencies

This cookbook depends on community versions of the following cookbooks:

* build-essential
* git
* nginx
* nodejs
* npm
* python
* redis

Additionally, it will install Ruby from default APT repos, and Foreman.

## Attributes

All keys below live under the <tt>cabot</tt> namespace. Eg. <tt>user</tt> is available as <tt>node[:cabot][:user]</tt>

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>user</tt></td>
    <td>String</td>
    <td>User to run Cabot as</td>
    <td><tt>cabot</tt></td>
  </tr>
  <tr>
    <td><tt>group</tt></td>
    <td>String</td>
    <td>Group to run Cabot as</td>
    <td><tt>cabot</tt></td>
  </tr>
  <tr>
    <td><tt>home_dir</tt></td>
    <td>String</td>
    <td>Home dir</td>
    <td><tt>/opt/cabot</tt></td>
  </tr>
  <tr>
    <td><tt>log_dir</tt></td>
    <td>String</td>
    <td>Log dir</td>
    <td><tt>/var/log/cabot</tt></td>
  </tr>
  <tr>
    <td><tt>repo_url</tt></td>
    <td>String</td>
    <td>GitHub repository to pull Cabot code from.</td>
    <td><tt>https://github.com/arachnys/cabot.git</tt></td>
  </tr>
  <tr>
    <td><tt>environment</tt></td>
    <td>String</td>
    <td>Sets the environment type (development or production). This sets whether to use SQLite or PostgreSQL databases.</td>
    <td><tt>production</tt></td>
  </tr>
  <tr>
    <td><tt>database_url</tt></td>
    <td>String</td>
    <td>Sets the Django database URL</td>
    <td><tt>sqlite:///cabot.db</tt></td>
  </tr>
  <tr>
    <td><tt>port</tt></td>
    <td>Integer</td>
    <td>Sets the port on which Cabot UI listens for requests</td>
    <td><tt>5000</tt></td>
  </tr>
  <tr>
    <td><tt>admin_email</tt></td>
    <td>String</td>
    <td>Emails sent from Django are sent to</td>
    <td><tt>you@example.com</tt></td>
  </tr>
  <tr>
    <td><tt>from_email</tt></td>
    <td>String</td>
    <td>From address to send emails with</td>
    <td><tt>cabot@example.com</tt></td>
  </tr>
  <tr>
    <td><tt>ical_url</tt></td>
    <td>String</td>
    <td>URL of calendar to synchronise rota with</td>
    <td><tt>http://www.google.com/calendar/ical/example.ics</tt></td>
  </tr>
  <tr>
    <td><tt>celery_broker_url</tt></td>
    <td>String</td>
    <td>Django setting</td>
    <td><tt>redis://:yourrediskey@localhost:6379/1</tt></td>
  </tr>
  <tr>
    <td><tt>django_secret_key</tt></td>
    <td>String</td>
    <td>Django setting</td>
    <td><tt>2FL6ORhHwr5eX34pP9mMugnIOd3jzVuT45f7w430Mt5PnEwbcJgma0q8zUXNZ68A</tt></td>
  </tr>
  <tr>
    <td><tt>graphite_api_url</tt></td>
    <td>String</td>
    <td>Hostname of Graphite server instance</td>
    <td><tt>http://graphite.example.com/</tt></td>
  </tr>
  <tr>
    <td><tt>graphite_username</tt></td>
    <td>String</td>
    <td>Username used to authenticate with Graphite</td>
    <td><tt>username</tt></td>
  </tr>
  <tr>
    <td><tt>graphite_password</tt></td>
    <td>String</td>
    <td>Password used to authenticate with Graphite</td>
    <td><tt>password</tt></td>
  </tr>
  <tr>
    <td><tt>hipchat_room_id</tt></td>
    <td>String</td>
    <td>Hipchat room ID (find it at https://hipchat.com/admin/rooms)</td>
    <td><tt>123456</tt></td>
  </tr>
  <tr>
    <td><tt>hipchat_api_key</tt></td>
    <td>String</td>
    <td>Hipchat API key (get one at https://hipchat.com/admin/api)</td>
    <td><tt>your_hipchat_api_key</tt></td>
  </tr>
  <tr>
    <td><tt>jenkins_api_url</tt></td>
    <td>String</td>
    <td>Jenkins server URL</td>
    <td><tt>https://jenkins.example.com/</tt></td>
  </tr>
  <tr>
    <td><tt>jenkins_username</tt></td>
    <td>String</td>
    <td>Username used to authenticate with Jenkins (optiona/td>
    <td><tt>username</tt></td>
  </tr>
  <tr>
    <td><tt>jenkins_password</tt></td>
    <td>String</td>
    <td>Password used to authenticate with Jenkins</td>
    <td><tt>password</tt></td>
  </tr>
  <tr>
    <td><tt>smtp_host</tt></td>
    <td>String</td>
    <td>SMTP hostname to use for sending emails</td>
    <td><tt>email-smtp.us-east-1.amazonaws.com</tt></td>
  </tr>
  <tr>
    <td><tt>smtp_username</tt></td>
    <td>String</td>
    <td>SMTP username</td>
    <td><tt>username</tt></td>
  </tr>
  <tr>
    <td><tt>smtp_password</tt></td>
    <td>String</td>
    <td>SMTP password</td>
    <td><tt>password</tt></td>
  </tr>
  <tr>
    <td><tt>smtp_port</tt></td>
    <td>String</td>
    <td>SMTP port</td>
    <td><tt>465</tt></td>
  </tr>
  <tr>
    <td><tt>twilio_account_sid</tt></td>
    <td>String</td>
    <td>Your Twilio account SID</td>
    <td><tt>your_twilio_account_sid</tt></td>
  </tr>
  <tr>
    <td><tt>twilio_auth_token</tt></td>
    <td>String</td>
    <td>Your Twilio auth token</td>
    <td><tt>your_twilio_auth_token</tt></td>
  </tr>
  <tr>
    <td><tt>twilio_outgoing_number</tt></td>
    <td>String</td>
    <td>Your Twilio number for outbound calls</td>
    <td><tt>+1234567890</tt></td>
  </tr>
  <tr>
    <td><tt>www_http_host</tt></td>
    <td>String</td>
    <td>Used for pointing links back in alerts, etc.</td>
    <td><tt>localhost</tt></td>
  </tr>
  <tr>
    <td><tt>www_scheme</tt></td>
    <td>String</td>
    <td>Which URL scheme to use (http or https)</td>
    <td><tt>http</tt></td>
  </tr>
  <tr>
    <td><tt>www_port</tt></td>
    <td>Integer</td>
    <td>Port to set reverse proxy on</td>
    <td><tt>80</tt></td>
  </tr>
  <tr>
    <td><tt>django_settings_module</tt></td>
    <td>String</td>
    <td>Django settings module name</td>
    <td><tt>cabot.settings</tt></td>
  </tr>
</table>

## Usage

### cabot::default

Include `cabot` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[cabot::default]"
  ]
}
```

### cabot::proxy

Include `cabot::proxy` in your node's `run_list` to have Nginx as a reverse proxy to Cabot on port 80.

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (i.e. `add-new-recipe`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request

## License and Authors

Author:: Rafael Fonseca (<rafael.magu@gmail.com>)


CHANGELOG.md
# 0.1.1

Updates to support with latest cabot from github master branch (tested on b84c524753d612b277c5287487a7e1da8c38c7d3)

# 0.1.0

Initial release of cabot

* Enhancements
  * an enhancement

* Bug Fixes
  * a bug fix


spec/nginx_spec.rb
require 'spec_helper'

describe 'cabot::proxy' do
  cached(:chef_run) { ChefSpec::Runner.new(log_level: :fatal).converge(described_recipe) }

  before do
    stub_command('which nginx').and_return('/usr/bin/nginx')
  end

  it 'should include nginx cookbook' do
    expect(chef_run).to include_recipe('nginx')
  end
end


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'
require 'chefspec/cacher'
require_relative 'support/matchers'

# ChefSpec::Coverage.start!

RSpec.configure do |config|
  config.platform = 'ubuntu'
  config.version = '12.04'
end


spec/default_spec.rb
require 'spec_helper'

describe 'cabot::default' do
  cached(:chef_run) { ChefSpec::Runner.new(log_level: :fatal).converge(described_recipe) }

  before do
    stub_command("npm -v 2>&1 | grep '1.4.20'").and_return(true)
  end

  context 'dependency setup' do
    %w(ubuntu git npm python redis build-essential).each do |cookbook|
      it "should include #{cookbook}" do
        expect(chef_run).to include_recipe(cookbook)
      end
    end

    %w(ruby1.9.1 libpq-dev).each do |pkg|
      it "should install #{pkg}" do
        expect(chef_run).to install_package(pkg)
      end
    end

    it 'should install foreman ruby gem' do
      expect(chef_run).to install_gem_package('foreman')
    end

    {
      'Django' => '1.4.10', 'PyJWT' => '0.1.2', 'South' => '0.7.6', 'amqp' => '1.4.5', 'anyjson' => '0.3.3',
      'argparse' => '1.2.1', 'billiard' => '3.3.0.13', 'celery' => '3.1.7', 'distribute' => '0.6.24',
      'dj-database-url' => '0.2.2', 'django-appconf' => '0.6', 'django-celery' => '3.1.1',
      'django-celery-with-redis' => '3.0', 'django-compressor' => '1.2', 'django-jsonify' => '0.2.1',
      'django-mptt' => '0.6.0', 'django-polymorphic' => '0.5.3', 'django-redis' => '1.4.5', 'django-smtp-ssl' => '1.0',
      'gunicorn' => '18.0', 'hiredis' => '0.1.1', 'httplib2' => '0.7.7', 'icalendar' => '3.2', 'kombu' => '3.0.8',
      'mock' => '1.0.1', 'psycopg2' => '2.5.1', 'pytz' => '2013.9', 'redis' => '2.9.0', 'requests' => '0.14.2',
      'six' => '1.5.1', 'twilio' => '3.4.1', 'wsgiref' => '0.1.2', 'python-dateutil' => '2.1'
    }.each do |mod, version|
      it "should install version #{version} of python module #{mod}" do
        expect(chef_run).to install_python_pip(mod).with(version: version)
      end
    end
  end

  context 'user setup' do
    it 'should create home dir manually' do
      expect(chef_run).to create_directory('/opt/cabot').with(owner: 'cabot', group: 'cabot')
    end

    it 'should create log dir manually' do
      expect(chef_run).to create_directory('/var/log/cabot').with(owner: 'cabot', group: 'cabot', mode: 0775)
    end

    it 'should create a cabot group' do
      expect(chef_run).to create_group('cabot')
    end

    it 'should create a cabot user' do
      expect(chef_run).to create_user('cabot').with(supports: { manage_home: false }, home: '/opt/cabot')
    end
  end

  context 'app deploy' do
    it 'should sync app code from repo' do
      expect(chef_run).to sync_git('/opt/cabot')
    end

    it 'should create an environment template for production' do
      expect(chef_run).to create_template('/opt/cabot/conf/production.env')
    end
  end
end


recipes/default.rb
#
# Cookbook Name:: cabot
# Recipe:: default
#
# Copyright (C) 2014 Rafael Fonseca
#
# MIT License
#

# dependency setup
%w(ubuntu git npm python redis build-essential).each do |cookbook|
  include_recipe cookbook
end

%w(ruby1.9.1 libpq-dev).each do |pkg|
  package pkg do
    action :install
  end
end

%w(coffee-script less).each do |pkg|
  npm_package pkg do
    action :install
  end
end

gem_package 'foreman' do
  action :install
end

{
  'Django' => '1.4.10', 'PyJWT' => '0.1.2', 'South' => '0.7.6', 'amqp' => '1.4.5', 'anyjson' => '0.3.3',
  'argparse' => '1.2.1', 'billiard' => '3.3.0.13', 'celery' => '3.1.7', 'distribute' => '0.6.24',
  'dj-database-url' => '0.2.2', 'django-appconf' => '0.6', 'django-celery' => '3.1.1',
  'django-celery-with-redis' => '3.0', 'django-compressor' => '1.2', 'django-jsonify' => '0.2.1',
  'django-mptt' => '0.6.0', 'django-polymorphic' => '0.5.3', 'django-redis' => '1.4.5', 'django-smtp-ssl' => '1.0',
  'gunicorn' => '18.0', 'hiredis' => '0.1.1', 'httplib2' => '0.7.7', 'icalendar' => '3.2', 'kombu' => '3.0.8',
  'mock' => '1.0.1', 'psycopg2' => '2.5.1', 'pytz' => '2013.9', 'redis' => '2.9.0', 'requests' => '0.14.2',
  'six' => '1.5.1', 'twilio' => '3.4.1', 'wsgiref' => '0.1.2', 'python-dateutil' => '2.1'
}.each do |mod, version|
  python_pip mod do
    action :install
    version version
  end
end

# user setup
group node[:cabot][:group]

user node[:cabot][:user] do
  supports manage_home: false
  home node[:cabot][:home_dir]
  group node[:cabot][:group]
  shell '/bin/bash'
end

directory node[:cabot][:home_dir] do
  action :create
  owner node[:cabot][:user]
  group node[:cabot][:group]
end

directory node[:cabot][:log_dir] do
  owner node[:cabot][:user]
  group node[:cabot][:group]
  mode 0775
end

# app deploy
git node[:cabot][:home_dir] do
  action :sync
  repository node[:cabot][:repo_url]
  user node[:cabot][:user]
  group node[:cabot][:group]
end

template "#{node[:cabot][:home_dir]}/conf/production.env" do
  action :create
  source 'production.env.erb'
  variables(
    debug: node[:cabot][:debug],
    database_url: node[:cabot][:database_url],
    port: node[:cabot][:port],
    virtualenv_dir: node[:cabot][:virtualenv_dir],
    admin_email: node[:cabot][:admin_email],
    from_email: node[:cabot][:from_email],
    ical_url: node[:cabot][:ical_url],
    celery_broker_url: node[:cabot][:celery_broker_url],
    django_secret_key: node[:cabot][:django_secret_key],
    graphite_api_url: node[:cabot][:graphite_api_url],
    graphite_username: node[:cabot][:graphite_username],
    graphite_password: node[:cabot][:graphite_password],
    hipchat_room_id: node[:cabot][:hipchat_room_id],
    hipchat_api_key: node[:cabot][:hipchat_api_key],
    jenkins_api_url: node[:cabot][:jenkins_api_url],
    jenkins_username: node[:cabot][:jenkins_username],
    jenkins_password: node[:cabot][:jenkins_password],
    smtp_host: node[:cabot][:smtp_host],
    smtp_username: node[:cabot][:smtp_username],
    smtp_password: node[:cabot][:smtp_password],
    smtp_port: node[:cabot][:smtp_port],
    twilio_account_sid: node[:cabot][:twilio_account_sid],
    twilio_auth_token: node[:cabot][:twilio_auth_token],
    twilio_outgoing_number: node[:cabot][:twilio_outgoing_number],
    www_http_host: node[:cabot][:www_http_host],
    www_scheme: node[:cabot][:www_scheme]
  )
  notifies :run, 'bash[run migrations]', :immediately
end

bash 'run migrations' do
  cwd node[:cabot][:home_dir]
  code <<-EOH
    foreman run python manage.py syncdb -e conf/#{node[:cabot][:environment]}.env
    foreman run python manage.py migrate cabotapp --noinput -e conf/#{node[:cabot][:environment]}.env
  EOH
  action :nothing
  notifies :run, 'bash[collect static assets]', :immediately
end

bash 'collect static assets' do
  cwd node[:cabot][:home_dir]
  code <<-EOH
    foreman run python manage.py collectstatic --noinput -e conf/#{node[:cabot][:environment]}.env
    foreman run python manage.py compress --force -e conf/#{node[:cabot][:environment]}.env
  EOH
  action :nothing
  notifies :run, 'bash[setup upstart]', :immediately
end

service 'cabot' do
  provider Chef::Provider::Service::Upstart
  action :nothing
end

bash 'setup upstart' do
  cwd node[:cabot][:home_dir]
  code <<-EOH
    foreman export upstart /etc/init -f #{node[:cabot][:home_dir]}/Procfile -e #{node[:cabot][:home_dir]}/conf/#{node[:cabot][:environment]}.env -u #{node[:cabot][:user]} -a cabot -t #{node[:cabot][:home_dir]}/upstart
  EOH
  action :nothing
  notifies :start, 'service[cabot]'
end
